pkg/scheduler/framework/preemption/preemption.go

```diff
@@ -410,25 +408,33 @@ func (ev *Evaluator) prepareCandidate(ctx context.Context, c Candidate, pod *v1.
 	return nil
 }
 
-// nodesWherePreemptionMightHelp returns a list of nodes with failed predicates
+// nodesWherePreemptionMightHelp returns a list of nodes with failed predicates (Unschedulable)
 // that may be satisfied by removing pods from the node.
-func nodesWherePreemptionMightHelp(nodes []*framework.NodeInfo, m framework.NodeToStatusMap) ([]*framework.NodeInfo, framework.NodeToStatusMap) {
+func nodesWherePreemptionMightHelp(allNodes []*framework.NodeInfo, nodeLister framework.NodeInfoLister, m framework.NodeToStatusReader) []*framework.NodeInfo {
 	var potentialNodes []*framework.NodeInfo
-	nodeStatuses := make(framework.NodeToStatusMap)
-	unresolvableStatus := framework.NewStatus(framework.UnschedulableAndUnresolvable, "Preemption is not helpful for scheduling")
-	for _, node := range nodes {
-		nodeName := node.Node().Name
-		// We only attempt preemption on nodes with status 'Unschedulable'. For
-		// diagnostic purposes, we propagate UnschedulableAndUnresolvable if either
-		// implied by absence in map or explicitly set.
-		status, ok := m[nodeName]
-		if status.Code() == framework.Unschedulable {
-			potentialNodes = append(potentialNodes, node)
-		} else if !ok || status.Code() == framework.UnschedulableAndUnresolvable {
-			nodeStatuses[nodeName] = unresolvableStatus
+
+	if m.AbsentNodesStatus().Code() == framework.Unschedulable {
+		if m.Len() == 0 {
+			return allNodes
+		}
+		for _, node := range allNodes {
```

Comment:
I think it would be hard to filter out the nodes using `ForEach` in this case faster than iterating through all the nodes. 

In this case, we want to filter the `Unschedulable` nodes when `Unschedulable` is also an `absentNodesStatus`, so we need to know all absent node names. I see only two solutions for this:
- My solution: Go through all the nodes and put into the result slice only these nodes, that are explicitly or implicitly `Unschedulable` (using `Get()` on `NodeToStatus`).
- Go through the nodes existing in `NodeToStatus` map and delete from the result these, that are not `Unschedulable`. The issue here is deleting the nodes, what would affect the performance significantly (transforming allNodes to set for faster delete ops and transform again to the final slice).